docs: add how-to guides for Go (#12)
canonical/ubuntu-for-developers-docs: docs/howto/go-setup.md

# How to set up development environment for Go on Ubuntu

Go is a popular language for backend web development,
microservices and CLI tools. This how-to guide outlines
how to install a Go distribution and set up a development
environment on Ubuntu.

## Install and set up Go

A Go installation includes the `go` command,
a compiler and other tools. 

There are multiple ways to install the Go toolchain on Ubuntu.

After using any installation method, confirm that
you have installed a specific version of Go with:

```
go version
```

### Package managers

Installing Go with a package manager is a convenient option.
This method does not require manually adding the `go` binary
to your PATH and future updates can be handled through the package manager.

#### Snap

Go can be installed as a snap:

```none
sudo snap install go --classic
```

> You can print a list of available Go releases with:

Author: I think most users will default to installing the latest version without checking available versions in advance. So I've kept the order but tried to make it more clear in the writing that listing the versions, installing a specific one, and switching versions, can be done in addition to the default install. See https://github.com/canonical/ubuntu-for-developers-docs/pull/12/commits/b7864f511aea31a53f5fd0ec981bc4baac04815f

```none
snap info go
```

To switch between release versions use `refresh`:

```none
sudo snap refresh go --channel=1.22/stable --classic
```

#### Debian package

Go can also be installed as a **Debian package**:

```
sudo apt install golang-go
```

> ```{note}

Author: Done in https://github.com/canonical/ubuntu-for-developers-docs/pull/12/commits/6b33f94089c89aa2fdcc8a37d0742f39af5740fb

:::

### Precompiled Go binaries

Precompiled Go binaries are available in a compressed format on the [release page](https://go.dev/dl/) of the official Go website.

1. Fetch a specific Go version with `wget`:

```none
wget https://go.dev/dl/go<version>.linux-amd64.tar.gz 
```

2. Extract the files into an appropriate directory:

```none
sudo tar -C /usr/local -xzf go<version>.linux-amd64.tar.gz
```

3. Add the `go` binary to your `$PATH` environment variable:

```none
export PATH=$PATH:/usr/local/go/bin
```

### Downloading and using multiple Go versions

Sometimes it may be necessary to run multiple Go versions
on the same machine.

The general command to install a specific version of Go is:

```none
sudo apt install golang-<version>
```

1. To install versions `1.21` and `1.23`, run:

```none
sudo apt install golang-1.21 golang-1.23
```

Each `go` binary will be installed in `/usr/lib/go-<version>/bin/`

2. Test the `go` binaries with:

```none
/usr/lib/go-1.21/bin/go version
/usr/lib/go-1.23/bin/go version
```

3. Optional: add binary to `$PATH`.

Make the name of the binary unique so that there is no
naming collision with any default `go` installation:

```none

> sudo mv /usr/lib/go-1.21/bin/go /usr/lib/go-1.21/bin/go1.21

Reviewer: Can we avoid this part. IIMO, modifying files in /usr should be avoided as they are supported to be managed by package manager.

Author: We can remove this -- I agree. Do you think it's sensible to keep running the distinct binaries by referencing full paths and aliases?

Reviewer: > Do you think it's sensible to keep running the distinct binaries by referencing full paths and aliases?

Maybe keep the full path. I'm not sure about the aliases, I haven't seen people use that before.

Author: Snap was restored as an installation method in https://github.com/canonical/ubuntu-for-developers-docs/pull/12/commits/8120dd629c39f7863efd0c09de7fa8a545c0804c and the section on installing multiple Go version was simplifed in https://github.com/canonical/ubuntu-for-developers-docs/pull/12/commits/2e0531a83e9327be4331ad6de51f5ffabddfe543 by, for example, removing the use of aliases.

```

To switch between release versions use `snap refresh`;
for example, to switch to the edge release of the latest Go version run:

```none
sudo snap refresh go --channel=latest/edge --classic
```

#### Debian package

To install Go using the default Ubuntu package manager:

```
sudo apt install golang-go
```

### Precompiled Go binaries

Precompiled Go binaries are available in a compressed format on the [release page](https://go.dev/dl/) of the official Go website and can be fetched with `wget`:

```none
wget https://go.dev/dl/go<version>.linux-amd64.tar.gz 
```

Extract the files into an appropriate directory:

```none
sudo tar -C /usr/local -xzf go<version>.linux-amd64.tar.gz
```

To use the `go` binary you need to add it to your `$PATH` environment variable.
Running the following command will modify the environmental variable for the current terminal session:

```none
export PATH=$PATH:/usr/local/go/bin
```

To persist the change and make `go` available in new terminal sessions and across reboots, 
append the above line to `$HOME/.profile` (or `/etc/profile`) and source the file:

```none
source $HOME/.profile
```

### Downloading and using multiple Go versions

> 

Reviewer: Multiple go versions are also available in Ubuntu archive. But users need to install the golang-1.XX-go package, and add /usr/lib/go-1.XX/bin to PATH.

Author: Should we include installs for Go and gopls from the Ubuntu archive in the guide also? If so, I can see about working them in.

Reviewer 2: @edibotopic Where possible, default to stuff that's in the Ubuntu repos. Only in cases where there's a specific need to mention both (e.g. if the latest version is not avail. from the archive, or if the package version is significantly different (missing proprietary piece?), add the alternative method, too.

Author: Thanks for the clarification @rkratky -- I will test this myself later today and add in the relevant sections.

Author: addressed in https://github.com/canonical/ubuntu-for-developers-docs/pull/12/commits/bb426225c80ab79cd0ffe87cc4cb8e1343b3d6fd

Add this to your `$PATH` to enable calling the version of Go that you have compiled.

```none
ALL TESTS PASSED
---
Installed Go for linux/amd64 in /home/<username>/build-go-from-source
Installed commands in /home/<username>/build-go-from-source/bin
*** You need to add /home/<username>/build-go-from-source/bin to your PATH.

```

3. Optional: To update Go to the latest version, `git pull` the latest changes and rebuild the binary with the `all.bash` script.

## Editing and debugging

Some of the most common Integrated Development Environments used for Go are:

- [Visual Studio Code](https://code.visualstudio.com/): a free editor with a dedicated [Go extension](https://marketplace.visualstudio.com/items?itemName=golang.Go) maintained by the Go Team
- [GoLand](https://www.jetbrains.com/go/): a paid editor developed specifically for the Go programming language

Both Visual Studio Code and GoLand rely on [Delve](https://github.com/go-delve/delve) for debugging.
Delve can also be installed as a standalone program and run [on the command line](https://github.com/go-delve/delve/blob/master/Documentation/cli/getting_started.md).

> An overview of basic Delve usage is included in our [how to develop with Go](./go-use.md) guide.

Reviewer: Should setting up delve move to this page? And delve is also available in Ubuntu archive.

Author: I think that might be better and the usage example can stay put.

Author: Done: https://github.com/canonical/ubuntu-for-developers-docs/pull/12/commits/0eb871c3ed9e6d868ed843a84da07cb55e0fddfb

> A Go language server [gopls](https://pkg.go.dev/golang.org/x/tools/gopls) is also actively maintained, which has helped ensure that Go is widely supported across many editors, including Emacs, (Neo)Vim and others. Install it by running:

Reviewer: gopls is also available in Ubuntu archive.

Author: https://github.com/canonical/ubuntu-for-developers-docs/pull/12/commits/bb426225c80ab79cd0ffe87cc4cb8e1343b3d6fd

```none
go install golang.org/x/tools/gopls@latest
```

## Cross-compilation

```{note}
For the basics of writing and testing a Hello World program in Go
see our [how to develop with Go](./go-use.md) guide.
```

Go has excellent cross-platform build capabilities.
To build a program called `hello.go`, containing valid Go code, run:

```none
go build hello.go
```

This builds a binary that can be run on your Ubuntu system.
Different target systems can be set for the compiler.

To set the environment for a Windows amd64 build, run:

```none
export GOOS=windows GOARCH=amd64
```

This causes `go build` to create a `hello.exe` binary that runs on Windows.

To test the (Linux) binary, execute it by running:

```none
./hello
```

### Building for multiple targets

For a full list of targets, run:

```none
go tool dist list
```

For this example, filter the output to Windows and Linux on `amd`:

```none
go tool dist list | grep 'amd' | grep -E 'windows|linux'
```

Create a makefile that automatically sets
the build environment and creates executable binaries for
both Windows and Linux platforms:

```make
EXE=heygo
WINDOWS=$(EXE)_win_amd64.exe

> LINUX=$(EXE)_lin_amd64

Reviewer: `_lin_` doesn't seem very common. Could we use `_linux_` here?

Author: https://github.com/canonical/ubuntu-for-developers-docs/pull/12/commits/ccf7bec4a1adbc9e433398cf631739160d40c001

After using any installation method, confirm that
you have installed a specific version of Go with:

```none
go version
```

### Package managers

Installing Go with a package manager is a convenient option.
This method does not require manually adding the `go` binary
to your `$PATH` and future updates can be handled through the package manager.

```{note}
Depending on the package manager used, the latest version of Go
may not be available.
See [Precompiled Go binaries](#precompiled-go-binaries) for instructions
on how to install the latest version of Go without a package manager.
```

#### Snap

To install the latest version of Go as a snap, run:

```none
sudo snap install go --classic
```

If you need to install a specific version, first enter the command:

```none
snap info go
```

> The output will indicate the snap channels that are available:

Reviewer 2: ```suggestion
The output indicates the snap channels that are available:
```

Author: fixed in https://github.com/canonical/ubuntu-for-developers-docs/pull/12/commits/102159689f21ca0b1d787203a13fd915d1ff0ca5

```none
...
...
channels:
  latest/stable:       1.23.2       2024-10-03 (10730)  68MB classic
  latest/candidate:    ↑                                     
  latest/beta:         ↑                                     
  latest/edge:         1.24-e705a2d 2024-08-07 (10683) 111MB classic
  1.23/stable:         1.23.2       2024-10-03 (10730)  68MB classic
  1.23/candidate:      1.23.2       2024-10-02 (10730)  68MB classic
  1.23/beta:           ↑                                     
  1.23/edge:           ↑                                     
  1.22/stable:         1.22.8       2024-10-03 (10719)  64MB classic
  1.22/candidate:      1.22.8       2024-10-02 (10719)  64MB classic
...
...
```

The stable version of Go 1.22 is installed with:

```none
sudo snap install go --channel=1.22/stable --classic
```

> To switch between release versions use `snap refresh`;
> for example, to switch to the edge release of the latest Go version run:

Reviewer 2: ```suggestion
To switch between release versions, use `snap refresh`;
for example, to switch to the edge release of the latest Go version, run:
```

Author: https://github.com/canonical/ubuntu-for-developers-docs/pull/12/commits/102159689f21ca0b1d787203a13fd915d1ff0ca5

go version
```

### Adding Go binary to path

Some steps in this guide require that Go is installed and
that the `go` command can be run from your terminal.

> If you are not using a package manager you will need

Reviewer 2: ```suggestion
If you are not using a package manager to install the tools, you need
```

to modify your `$PATH` manually.

To temporarily add Go to your `$PATH`, run:

> 	export PATH=$PATH:/path/to/your/go/bin

Reviewer 2: ```suggestion
```none
export PATH=$PATH:/path/to/your/go/bin
```
```

Reviewer 2: Hmm, misformatted. Missing three backticks at the bottom.

Author: Ah, this section was originally a note block and the indents were used for the code. Thanks for spotting.

To persist the change and make go available in new terminal sessions and across reboots, 
append the above line to `$HOME/.profile` (or `/etc/profile`) and source the file:

> 	source $HOME/.profile

Reviewer 2: ```suggestion
```none
source $HOME/.profile
``` 
```

Reviewer 2: Same.

> The files to modify and commands to use when modifying the `$PATH` may vary
> depending on your shell environment.

Reviewer 2: ```suggestion
:::{note}
The files to modify and commands to use when modifying the `$PATH` may vary
depending on your shell environment.
:::
```

```

To use the `go` binary you need to add it to your `$PATH` environment variable.

```none
export PATH=$PATH:/usr/local/go/bin
```

### Downloading and using multiple Go versions

Sometimes it may be necessary to run multiple Go versions
on the same machine.

#### Ubuntu archive

When installing a specific version of Go the command is:

```none
sudo apt install golang-<version>
```

1. To install versions `1.21` and `1.23`:

```none
sudo apt install golang-1.21 golang-1.23
```

2. To temporarily add a specific version to your `$PATH`:

```none
export PATH=$PATH:/usr/lib/go-1.21/bin
```

3. To check the version:

```none
go version go1.21.9 linux/amd64
```

> 4. Optional: if you want to easily switch between versions use aliases.

Reviewer 2: ```suggestion
4. Optional: to easily switch between versions, use aliases.
```

For example, in a `.bashrc` file:

```none
alias go1.21='/usr/lib/go-1.21/bin/go'
alias go1.23='/usr/lib/go-1.23/bin/go'
```

These can now be called separately:

```none
go1.21 version
```

```none
go1.23 version
```

> If you have a default installation of Go installed with `apt install golang-go`

Reviewer 2: ```suggestion
If you have a default installation of Go installed with `apt install golang-go`,
```

## Editing and debugging

### Integrated Development Environments and editors

Some of the most common Integrated Development Environments (IDEs) used for Go are:

- [Visual Studio Code](https://code.visualstudio.com/): a free editor with a dedicated [Go extension](https://marketplace.visualstudio.com/items?itemName=golang.Go) maintained by the Go Team
- [GoLand](https://www.jetbrains.com/go/): a paid editor developed specifically for the Go programming language

A Go language server [gopls](https://pkg.go.dev/golang.org/x/tools/gopls) is also actively maintained, which has helped ensure that Go is widely supported across many editors, including Emacs, (Neo)Vim and others. 

You can install gopls from the Ubuntu archive:

```none
sudo apt install gopls
```

Alternatively, if `go` is in your `$PATH` you can use `go install`:

```none
go install golang.org/x/tools/gopls@latest
```

### Debugging

IDEs used for Go development commonly rely on [Delve](https://github.com/go-delve/delve) for debugging.
Delve can also be installed as a standalone program and run [on the command line](https://github.com/go-delve/delve/blob/master/Documentation/cli/getting_started.md).

> To install Delve run:

Reviewer 2: ```suggestion
To install Delve, run:
```

```none
sudo apt install delve
```

> If Go is installed and in your `$PATH` then Delve can also be installed with `go install`:
> 
> ```none
> go install github.com/go-delve/delve/cmd/dlv@latest
> ```

Reviewer 2: Do we need this? Is there any benefit to installing this way? If not, all non-apt installation instructions can be removed.
```suggestion
If Go is installed and in your `$PATH`, then Delve can also be installed with `go install`:

```none
go install github.com/go-delve/delve/cmd/dlv@latest
```
```

Author: It might be easier to install the latest version with `go install` and it's commonly used.
I agree that we should default to apt and in a revised version will check if there can perhaps be a general comment about `go install` being an option, while not emphasising it throughout.

Author: **Update** @rkratky @zhsj 

Based on the comment from @rkratky above and some additional discussions about same, in https://github.com/canonical/ubuntu-for-developers-docs/pull/12/commits/75160dfc71268002a1a576cd4bafe6e6f5867f4b I have increased the emphasis placed on `apt` as the installation method. I have removed the instructions for building Go from source and installing Go as a snap. These can be added at a later date if there is a compelling reason to do so. I have also added some more detail around installing multiple versions using `apt`, specifically around avoiding having multiple `go` binaries in PATH.

I have also added a generic note about how `go install` can be used for many Go tools and indicated that the two IDEs highlighted can be installed as snaps.

To accommodate these changes I had to restructure the installation guide a little.

Reviewer: installing go as snap should be kept, as snap is also an important feature for ubuntu.

Author: Thanks. In this case, is there a reason we could provide for why a Go developer might choose a snap over the apt install?

Reviewer 2: @zhsj Is the reason for including the snap install just to be nice to snap, or can we cite some other advantage (or distinguishing feature) of the snap version? If it's exactly the same as the archive version, then I'd suggest adding just a small note to the tune of "A snap package of Go is also available for installing using `snap install ...`."

Reviewer: snap provides all go versions, regardless of the ubuntu release version. IMO it should be preferred over the `apt` version. However snap comes with its own shortage, for example it can't used in docker container.

Author: Very useful background @zhsj . I will put the Snap install back in the doc and indicate that this is a viable option with good support for different releases. I can mention that the choice may be influenced by the use case, e.g., docker and availability of versions. I think we should avoid indicating a preference at this point.

Does that sound OK?

@rkratky and I discussed that it might be useful in future to have some general top-level discussion in the docs on package management, and not have the discussion completely tied to each toolchain.

Reviewer 2: > IMO it should be preferred over the apt version.

We need to have (at some point; not now) a discussion about our approach to this because our message should be consistent across the board.

For now, let's go with what @edibotopic is suggesting above.

Reviewer: Yes, agree. (Sorry for missing the last comment)
